File: android/src/main/java/com/margelo/nitro/blesdk/BleSdk.kt
package com.margelo.nitro.blesdk

import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.blesdk.bluetooth.facade.BleFacade

@DoNotStrip
class BleSdk : HybridBleSdkSpec() {
  private var bleFacade: BleFacade? = null

  private fun getBleFacade(): BleFacade {
    bleFacade?.let { facade ->
      return facade
    }

    val reactContext = NitroModules.applicationContext
      ?: throw IllegalStateException(
        "NitroModules.applicationContext is not available. Make sure Nitro is installed before using BleSdk."
      )

    return BleFacade(reactContext.applicationContext).also { facade ->
      bleFacade = facade
    }
  }

  override fun isBluetoothSupported(): Boolean {
    return getBleFacade().isBluetoothSupported()
  }

  override fun isBluetoothEnabled(): Boolean {
    return getBleFacade().isBluetoothEnabled()
  }

  override fun isScanning(): Boolean {
    return getBleFacade().isScanning()
  }

  override fun getRequiredPermissions(): Array<String> {
    return getBleFacade().getRequiredPermissions()
  }

  override fun hasRequiredPermissions(): Boolean {
    return getBleFacade().hasRequiredPermissions()
  }

  override fun startScan(serviceUuid: String?, deviceName: String?, scanMode: Double?) {
    getBleFacade().startScan(serviceUuid, deviceName, scanMode)
  }

  override fun stopScan() {
    getBleFacade().stopScan()
  }

  override fun pairDevice(address: String): Boolean {
    return getBleFacade().pairDevice(address)
  }

  override fun unpairDevice(address: String): Boolean {
    return getBleFacade().unpairDevice(address)
  }

  override fun connectDevice(address: String): Boolean {
    return getBleFacade().connectDevice(address)
  }

  override fun disconnectDevice(address: String): Boolean {
    return getBleFacade().disconnectDevice(address)
  }

  override fun clearDiscoveredDevices() {
    getBleFacade().clearDiscoveredDevices()
  }

  override fun getDiscoveredDevices(): Array<BleDevice> {
    return getBleFacade()
      .getDiscoveredDevices()
      .map { device -> device.toBleDevice() }
      .toTypedArray()
  }

  override fun dispose() {
    bleFacade?.dispose()
    bleFacade = null
    super.dispose()
  }
}

File: ios/BleSdk.swift
import Foundation

private enum BleSdkError: LocalizedError {
    case unsupportedPlatform

    var errorDescription: String? {
        return "BLE scan, pair, and connection flow is currently implemented only on Android."
    }
}

class BleSdk: HybridBleSdkSpec {
    private var discoveredDevices: [BleDevice] = []
    private var scanning = false

    public func isBluetoothSupported() throws -> Bool {
        return false
    }

    public func isBluetoothEnabled() throws -> Bool {
        return false
    }

    public func isScanning() throws -> Bool {
        return scanning
    }

    public func getRequiredPermissions() throws -> [String] {
        return []
    }

    public func hasRequiredPermissions() throws -> Bool {
        return true
    }

    public func startScan(
        serviceUuid: String?,
        deviceName: String?,
        scanMode: Double?
    ) throws {
        scanning = false
        throw BleSdkError.unsupportedPlatform
    }

    public func stopScan() throws {
        scanning = false
    }

    public func pairDevice(address: String) throws -> Bool {
        throw BleSdkError.unsupportedPlatform
    }

    public func unpairDevice(address: String) throws -> Bool {
        throw BleSdkError.unsupportedPlatform
    }

    public func connectDevice(address: String) throws -> Bool {
        throw BleSdkError.unsupportedPlatform
    }

    public func disconnectDevice(address: String) throws -> Bool {
        throw BleSdkError.unsupportedPlatform
    }

    public func clearDiscoveredDevices() throws {
        discoveredDevices.removeAll()
    }

    public func getDiscoveredDevices() throws -> [BleDevice] {
        return discoveredDevices
    }
}
